mypy/server/update.py

```diff
@@ -1027,6 +1027,10 @@ def key(node: FineGrainedDeferredNode) -> int:
         if graph[module_id].type_checker().check_second_pass():
             more = True
 
+    graph[module_id].detect_possibly_undefined_vars()
```

Comment:
**Question:** I haven't added a call to `finish_passes()` here. I'm not sure what it does... should it be here?

Reply:
I don't think so, since `finish_passes` seems to only perform things related to non-fine-grained builds.

I think that there may be another issue here: `detect_possible_undefined_vars()` is somewhat expensive, as it runs over the entire file. In this function it would be better to only analyze the nodes that are being processed. For example, we could pass the set of targets to `detect_possibly_undefined_vars`, and it would only catch errors in these targets. I'm not sure if processing the entire file causes correctness issues, but plausibly it could.